objects_test.go

```diff
@@ -108,6 +115,24 @@ func (s *ObjectsSuite) TestBlobHash(c *C) {
 	c.Assert(string(data), Equals, "FOO")
 }
 
+func (s *ObjectsSuite) TestBlobDecodeEncodeIdempotent(c *C) {
+	var objects []*core.MemoryObject
+	obj := &core.MemoryObject{}
+	obj.Write([]byte("foo"))
+	obj.SetType(core.BlobObject)
+	obj.Hash()
```

Comment:
We have to encode a Blob into a core.Object. Blob implementation contains an inner core.Object, but we're not guaranteed that this inner core.Object is of the same type as the one we want to encode to.

If we want to change it, we'd probably need to simplify the whole thing and make core.MemoryObject the only representation of object and/or make .Encode() create its own core.Object, instead of receiving one.